lago/export.py

```diff
@@ -0,0 +1,239 @@
+
+    DiskExportManger is the base class of specific ExportManagers.
+    Each specific ExportManger is responsible on the export process of an
+    image with a specific type (e.g template, file...)
+
+    Attributes:
+        src (str): Path to the image that should be exported
+        name (str): The name of the exported disk
+        dst (str): The absolute path of the exported disk
+        disk_type (str): The type of the image e.g template, file, empty...
+        disk (dict): Disk attributes (of the disk that should be exported)
+         as found in workdir/current/virt/VM-NAME
+        exported_metadata(dict): A copy of the source disk metadata, this
+         dict should be updated with new values during the export process.
+        do_compress(bool): If true, apply compression to the exported disk.
+    """
+
+    __metaclass__ = ABCMeta
+
+    def __init__(self, dst, disk_type, disk, do_compress=False):
+        """
+        Args:
+            dst (str): The absolute path of the exported disk
+            disk_type (str): The type of the image
+             e.g template, file, empty...
+            disk (dict): Disk attributes as found in
+             workdir/current/virt/VM-NAME
+            do_compress(bool): If true, apply compression to the
+             exported disk.
+        """
+        self.src = path.expandvars(disk['path'])
+        self.name = path.basename(self.src)
+        self.dst = path.join(path.expandvars(dst), self.name)
+        self.dist_type = disk_type
+        self.disk = disk
+        self.exported_metadata = copy.deepcopy(disk['metadata'])
+        self.do_compress = do_compress
+
+    @staticmethod
+    def get_instance_by_type(dst, disk, do_compress=False, *args, **kwargs):
+        """
+        Args:
+            dst (str): The path of the new exported disk.
+             can contain env variables.
+            disk (dict): Disk attributes
+             (of the disk that should be exported) as
+             found in workdir/current/virt/VM-NAME
+            do_compress(bool): If true, apply compression to the exported disk.
+        Returns
+            An instance of a subclass of DiskExportManager which
+             matches the disk type.
+        """
+        disk_type = disk['type']
+        cls = HANDLERS.get(disk_type)
+        if not cls:
+            raise NotImplementedError(
+                'Export is not supported for disk type {}'.format(disk_type)
+            )
+        return cls(dst, disk_type, disk, do_compress, *args, **kwargs)
+
+    @abstractmethod
+    def export(self):
+        """
+        This method will handle the export process and should
+         implemented in
+        each subclass.
+        """
+        raise NotImplementedError
+
+    def copy(self):
+        """
+        Copy the disk using cp in order to preserves the 'sparse'
+        structure of the file
+        """
+        with LogTask('Copying disk'):
+            utils.cp(self.src, self.dst)
```

Comment:
Does that copy with sparseness? If not (and even if it does), we should probably run 'virt-sparsify' on the image first. It takes few seconds, and reduces nicely (by up to several hundreds of megs) our disks. virt-sparsify is better than plain 'cp' with sparseness preserveness, as it 'compacts' the image as well - sparsifies marked-as-deleted-but-not-deleted blocks.